drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -529,42 +386,70 @@ Error AudioDriverWASAPI::init_input_device(bool p_reinit) {
 	return OK;
 }
 
-Error AudioDriverWASAPI::audio_device_finish(AudioDeviceWASAPI *p_device) {
-	if (p_device->active.is_set()) {
-		if (p_device->audio_client) {
-			p_device->audio_client->Stop();
-		}
-		p_device->active.clear();
+void AudioDriverWASAPI::audio_device_finish(AudioDeviceWASAPI *p_device) {
+	if (p_device->audio_client) {
+		p_device->audio_client->Stop();
+	}
+
+	// Only used for capture.
+	p_device->active.clear();
+
+	if (p_device->feed_event) {
+		CloseHandle(p_device->feed_event);
+		p_device->feed_event = nullptr;
 	}
 
 	SAFE_RELEASE(p_device->audio_client)
 	SAFE_RELEASE(p_device->render_client)
 	SAFE_RELEASE(p_device->capture_client)
+}
 
-	return OK;
+void AudioDriverWASAPI::finish_output_device() {
+	audio_device_finish(&audio_output);
 }
 
-Error AudioDriverWASAPI::finish_output_device() {
-	return audio_device_finish(&audio_output);
+void AudioDriverWASAPI::finish_input_device() {
+	audio_device_finish(&audio_input);
 }
 
-Error AudioDriverWASAPI::finish_input_device() {
-	return audio_device_finish(&audio_input);
+// Only to be called in the render thread after init has already happened!
+Error AudioDriverWASAPI::remake_output_device() {
+	finish_output_device();
+
+	// At this point the wake event will be signaled. If the remake works, then the event will become signaled again and everything works.
+	// If the remake fails, then it will remain unsignaled and the render thread goes back to waiting for a working output device.
+
+	Error err = init_output_device(true); // This will fail if the device isn't in the system anymore.
+	if (err == OK) {
+		start(); // Will signal the wake event.
+	} else {
+		ResetEvent(render_wake); // Not active anymore. Put to waiting.
+	}
+
+	return err;
 }
 
 Error AudioDriverWASAPI::init() {
 	mix_rate = _get_configured_mix_rate();
 
-	target_latency_ms = Engine::get_singleton()->get_audio_output_latency();
-
+	// This below is cryptic but it is assumed that if we fail to initialize now, then maybe a new working device will be inserted to the system later.
 	Error err = init_output_device();
 	if (err != OK) {
 		ERR_PRINT("WASAPI: init_output_device error");
 	}
 
+	// The initial state for this event must be unsignaled!
+	// Literally impossible for this call to fail. You be the judge if you need to check the return value of this.
+	// It is however fundamental to the logic for this to exist.
+	render_wake = CreateEventW(nullptr, TRUE, FALSE, nullptr);
+	ERR_FAIL_COND_V_MSG(render_wake == nullptr, ERR_CANT_OPEN, "WASAPI: Could not create render wake handle with error " + String::num(GetLastError()));
```

Comment:
```suggestion
	ERR_FAIL_NULL_V_MSG(render_wake, ERR_CANT_OPEN, "WASAPI: Could not create render wake handle with error: " + String::num(GetLastError()) + ".");
```

Reply:
The condition here is correct. The suggested change would cause success to fail.

Reply:
Same here, you are misunderstanding the check